support-models/src/main/scala/com/gu/support/workers/states/SendThankYouEmailState.scala

```diff
@@ -2,36 +2,139 @@ package com.gu.support.workers.states
+    product: GuardianWeekly,
+    giftRecipient: Option[WeeklyGiftRecipient],
+    purchaseInfo: PurchaseInfo,
+    firstDeliveryDate: LocalDate,
+  ) extends ProductTypeCreated with PurchaseCreated
+
+  val circeDiscriminator = new CirceDiscriminator("productTypeCreatedType")
+
+  private val discriminatorContribution = "Contribution"
+  private val discriminatorDigitalSubscriptionDirectPurchase = "DigitalSubscriptionDirectPurchase"
+  private val discriminatorDigitalSubscriptionGiftPurchase = "DigitalSubscriptionGiftPurchase"
+  private val discriminatorDigitalSubscriptionCorporateRedemption = "DigitalSubscriptionCorporateRedemption"
+  private val discriminatorDigitalSubscriptionGiftRedemption = "DigitalSubscriptionGiftRedemption"
+  private val discriminatorPaper = "Paper"
+  private val discriminatorGuardianWeekly = "GuardianWeekly"
+
+  implicit val c1 = circeDiscriminator.discriminatedCodec[ContributionCreated](discriminatorContribution)
+  implicit val c2 = circeDiscriminator.discriminatedCodec[DigitalSubscriptionDirectPurchaseCreated](discriminatorDigitalSubscriptionDirectPurchase)
+  implicit val c3 = circeDiscriminator.discriminatedCodec[DigitalSubscriptionGiftPurchaseCreated](discriminatorDigitalSubscriptionGiftPurchase)
+  implicit val c4 = circeDiscriminator.discriminatedCodec[DigitalSubscriptionCorporateRedemptionCreated](discriminatorDigitalSubscriptionCorporateRedemption)
+  implicit val c5 = circeDiscriminator.discriminatedCodec[DigitalSubscriptionGiftRedemptionCreated](discriminatorDigitalSubscriptionGiftRedemption)
+  implicit val c6 = circeDiscriminator.discriminatedCodec[PaperCreated](discriminatorPaper)
+  implicit val c7 = circeDiscriminator.discriminatedCodec[GuardianWeeklyCreated](discriminatorGuardianWeekly)
+
+  implicit val encoder: Encoder[ProductTypeCreated] = Encoder.instance {
+    case c: ContributionCreated => c.asJson
+    case c: DigitalSubscriptionDirectPurchaseCreated => c.asJson
+    case c: DigitalSubscriptionGiftPurchaseCreated => c.asJson
+    case c: DigitalSubscriptionCorporateRedemptionCreated => c.asJson
+    case c: DigitalSubscriptionGiftRedemptionCreated => c.asJson
+    case c: PaperCreated => c.asJson
+    case c: GuardianWeeklyCreated => c.asJson
+  }
+
+  implicit val decoder =
+    circeDiscriminator.decode[ProductTypeCreated](Map(
+      discriminatorContribution -> Decoder[ContributionCreated],
+      discriminatorDigitalSubscriptionDirectPurchase -> Decoder[DigitalSubscriptionDirectPurchaseCreated],
+      discriminatorDigitalSubscriptionGiftPurchase -> Decoder[DigitalSubscriptionGiftPurchaseCreated],
+      discriminatorDigitalSubscriptionCorporateRedemption -> Decoder[DigitalSubscriptionCorporateRedemptionCreated],
+      discriminatorDigitalSubscriptionGiftRedemption -> Decoder[DigitalSubscriptionGiftRedemptionCreated],
+      discriminatorPaper -> Decoder[PaperCreated],
+      discriminatorGuardianWeekly -> Decoder[GuardianWeeklyCreated],
+    ))
+}
```

Comment:
there is a bit much copy and paste here for my liking.  we could get the compiler to wire some of it up correctly for us, along these lines:
(but if we might switch to circe-extras (which uses macros and does more automatically) then it may not be worth playing too much.)
```scala
  val circeDiscriminator = new CirceDiscriminator("productTypeCreatedType")

  implicit val discriminatorContribution = Discriminator[ContributionCreated]("Contribution")
  implicit val discriminatorDigitalSubscriptionDirectPurchase = Discriminator[DigitalSubscriptionDirectPurchaseCreated]("DigitalSubscriptionDirectPurchase")
  implicit val discriminatorDigitalSubscriptionGiftPurchase = Discriminator[DigitalSubscriptionDirectPurchaseCreated]("DigitalSubscriptionGiftPurchase")
  implicit val discriminatorDigitalSubscriptionCorporateRedemption = Discriminator[DigitalSubscriptionCorporateRedemptionCreated]("DigitalSubscriptionCorporateRedemption")
  implicit val discriminatorDigitalSubscriptionGiftRedemption = Discriminator[DigitalSubscriptionGiftRedemptionCreated]("DigitalSubscriptionGiftRedemption")
  implicit val discriminatorPaper = Discriminator[PaperCreated]("Paper")
  implicit val discriminatorGuardianWeekly = Discriminator[GuardianWeeklyCreated]("GuardianWeekly")

  implicit val encoder: Encoder[ProductTypeCreated] = Encoder.instance {
    case c: ContributionCreated => c.asJson
    case c: DigitalSubscriptionDirectPurchaseCreated => c.asJson
    case c: DigitalSubscriptionGiftPurchaseCreated => c.asJson
    case c: DigitalSubscriptionCorporateRedemptionCreated => c.asJson
    case c: DigitalSubscriptionGiftRedemptionCreated => c.asJson
    case c: PaperCreated => c.asJson
    case c: GuardianWeeklyCreated => c.asJson
  }

  implicit val decoder =
    circeDiscriminator.decode[ProductTypeCreated](List(
      DiscriminatedDecoder[ContributionCreated],
      DiscriminatedDecoder[DigitalSubscriptionDirectPurchaseCreated],
      DiscriminatedDecoder[DigitalSubscriptionGiftPurchaseCreated],
      DiscriminatedDecoder[DigitalSubscriptionCorporateRedemptionCreated],
      DiscriminatedDecoder[DigitalSubscriptionGiftRedemptionCreated],
      DiscriminatedDecoder[PaperCreated],
      DiscriminatedDecoder[GuardianWeeklyCreated],
    ))
}
```

Reply:
actually now using DiscriminatedType is much simpler than even this example